### PR TITLE
C++ Average Pool Module

### PR DESCRIPTION
**Pull request by ShahriarRezghi**
This PR adds Average Pool module to C++ front-end.

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/avgpool.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+/// Options for a `D`-dimensional avgpool module.
+template <size_t D>
+struct AvgPoolOptions {
+  AvgPoolOptions(ExpandingArray<D> kernel_size)
+      : kernel_size_(kernel_size), stride_(kernel_size) {}
+
+  /// the size of the window to take a average over
+  TORCH_ARG(ExpandingArray<D>, kernel_size);
+
+  /// the stride of the window. Default value is `kernel_size
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/avgpool.h]: nit:
```suggestion
  /// the stride of the window. Default value is `kernel_size`
```

Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/avgpool.h]: nit:
```suggestion
  /// the stride of the window. Default value is `kernel_size`
```

File: torch/csrc/api/include/torch/nn/modules/avgpool.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+namespace torch {
+namespace nn {
+
+/// Options for a `D`-dimensional avgpool module.
+template <size_t D>
+struct AvgPoolOptions {
+  AvgPoolOptions(ExpandingArray<D> kernel_size)
+      : kernel_size_(kernel_size), stride_(kernel_size) {}
+
+  /// the size of the window to take a average over
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/avgpool.h]: nit:
```suggestion
  /// the size of the window to take an average over
```

File: test/cpp/api/modules.cpp [hunk 3]
```diff
@@ -103,6 +104,58 @@ TEST_F(ModulesTest, Conv3d) {
...
 
+TEST_F(ModulesTest, AvgPool1d) {
+  AvgPool1d model(AvgPool1dOptions(3).stride(2));
+  auto x = torch::ones({1, 1, 5}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_TRUE(torch::allclose(y, torch::ones({1, 1 ,2})));
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: nit:
```suggestion
  ASSERT_TRUE(torch::allclose(y, torch::ones({1, 1, 2})));
```

File: test/cpp/api/modules.cpp [hunk 4]
```diff
@@ -103,6 +104,58 @@ TEST_F(ModulesTest, Conv3d) {
...
+
+TEST_F(ModulesTest, AvgPool2dEven) {
+  AvgPool2d model(AvgPool2dOptions(3).stride(2));
+  auto x = torch::ones({2, 5, 5}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_TRUE(torch::allclose(y, torch::ones({2, 2 ,2})));
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: nit:
```suggestion
  ASSERT_TRUE(torch::allclose(y, torch::ones({2, 2, 2})));
```

**Review by yf225 — COMMENTED** [review]
Thanks a lot @ShahriarSS ! I left some very minor nit comments.

File: caffe2/CMakeLists.txt [hunk 5]
```diff
@@ -509,6 +509,7 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       ${TORCH_SRC_DIR}/csrc/api/src/jit.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/init.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/module.cpp
+      ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/avgpool.cpp
```
Comment (yf225) [inline caffe2/CMakeLists.txt]: @ShahriarSS I am thinking we might need to add the `avgpool.cpp` entry into https://github.com/pytorch/pytorch/blob/13292ec3c7f47228b93dbef1ca750b66b4b8e0af/tools/build_variables.py#L178-L205 similar to the other C++ module PRs

File: torch/csrc/api/include/torch/nn/modules/avgpool.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  TORCH_ARG(ExpandingArray<D>, stride);
+
+  /// implicit zero padding to be added on both sides
+  TORCH_ARG(ExpandingArray<D>, padding) = 0;
+
+  /// when True, will use `ceil` instead of `floor` to compute the output shape
+  TORCH_ARG(bool, ceil_mode) = false;
+
+  /// when True, will include the zero-padding in the averaging calculation
+  TORCH_ARG(bool, count_include_pad) = true;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/avgpool.h]: @ShahriarSS Is there a specific reason we don't have `divisor_override` as in the Python version?

Comment (ShahriarSS) [reply torch/csrc/api/include/torch/nn/modules/avgpool.h]: @yf225 Sorry I based the inputs on ```AvgPool1d``` args and didn't see that option.

**Review by yf225 — APPROVED** [review]
Thanks a lot @ShahriarSS !

**Pull request closed without merging**